File: construct_bst_in_pre.cpp
// Time Complexity : O(n)
// Space Complexity :  O(n*n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Couldn't solve it properly. Took reference from lecture

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        
        if(preorder.size() == 0)
            return NULL;
        
        int root_val = preorder[0];
        int root_index = 0;
        TreeNode* root = new TreeNode;
        root->val = root_val;

        // Find the index of root element        
        for(int i = 0; i < inorder.size(); i++)
        {
            if(inorder[i] == root->val)
                root_index = i;
        }
        
        vector<int> pre_left;
        vector<int> pre_right;
        vector<int> in_left;
        vector<int> in_right;
        
        // Divide the Trees depending on the root element
        copy(preorder.begin()+1, preorder.begin()+root_index+1, back_inserter(pre_left));
        copy(preorder.begin()+root_index+1, preorder.begin()+preorder.size(), back_inserter(pre_right));
        copy(inorder.begin(), inorder.begin()+root_index, back_inserter(in_left));
        copy(inorder.begin()+root_index+1, inorder.begin()+inorder.size(), back_inserter(in_right));
        
        // Build the subtrees
        root->left = buildTree(pre_left, in_left);
        root->right = buildTree(pre_right, in_right);
        
        return root;
    }
};

File: src/ConstructBSTInPre.java
// Time Complexity : O(N)
// Space Complexity : O(N) (Recursion stack + hashmap)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

import java.util.HashMap;
import java.util.Map;

/**
 * Put inorder indices in a map so that finding the right index will be O(1).
 * Then take a global index to find the preorder index position. Take this
 * indexed val as the root and for root with left as 0 and right n-1. Now,
 * for root.left update right as index -1 from map of the current val and for
 * root.right update left as index +1 from map of the current val.
 *
 */
class Solution {
    int preOrderIndex = 0;
    Map<Integer, Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for(int i=0;i<inorder.length;i++)
            map.put(inorder[i], i);
        return helper(preorder, 0, preorder.length - 1);
    }

    public TreeNode helper(int[] preorder, int left, int right) {
        if(left > right)
            return null;
        
        int val = preorder[preOrderIndex++];
        TreeNode root = new TreeNode(val);
        root.left = helper(preorder, left, map.get(val) - 1);
        root.right = helper(preorder, map.get(val) + 1, right);
        return root;
    }
}}
